raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -395,31 +393,35 @@ private void resetConnections() {
         requestManager.resetAll();
     }
 
-    private void onBecomeLeader(long currentTimeMs) {
-        LeaderState state = quorum.leaderStateOrThrow();
+    private void onBecomeLeader(long currentTimeMs) throws IOException {
+        long endOffset = log.endOffset().offset;
+
+        // Add 1 to the offset that the accumulator tracks since appendLeaderChangeMessage 
+        // will write a record from the new leader's epoch to advance the high watermark below
+        BatchAccumulator<T> accumulator = new BatchAccumulator<>(
+            quorum.epoch(),
+            endOffset + 1,
```

Comment:
Hmm.. This is a little strange. It would be nice if we could refactor this so that we can write the LeaderChange message through LeaderState. This is a little bit tricky, but it should be doable. Let me try to put this together.